manager/orchestrator/update/updater.go

```diff
@@ -644,3 +658,10 @@ func (u *Updater) completeUpdate(ctx context.Context, serviceID string) {
 		log.G(ctx).WithError(err).Errorf("failed to mark update of service %s complete", serviceID)
 	}
 }
+
+func (u *Updater) annotationsUpdated(service *api.Service) bool {
```

Comment:
This was my naive way to double-check that no tasks had to be updated; I couldn't find a canonical way to determine if it's a spec-only change, or if tasks (containers) had to be updated; perhaps there is somewhere

Reply:
what's the problem with trusting the `dirtySlots` check? (keeping in mind that I haven't expanded that part of the code yet)

Reply:
IIUC;

- The `dirtySlots` check checks if any of the containers (tasks) have to be updated, so that the reconciliation loop has to check for that to finish
- If `len(dirtySlots) == 0`, no containers have to be updated

At that point, there's three possibilities;

1. `service.UpdateStatus == updating or rollback_started` - this means a reconciliation loop was started and has now finished (so `completeUpdate`)
2. `service.UpdateStatus == nil`, but nothing changed (`annotationsUpdated == false`); apparently there was nothing to update; don't update the `Service.UpdateStatus`, because nothing happened
3. `service.UpdateStatus == nil` and annotations changed; this means we actually updated something, and are now done; run `completeUpdate` to set the `Service.UpdateStatus`


Perhaps we don't have to take `2.` into account, and always call `completeUpdate` update the `UpdateStatus` once we get at this point; I wasn't sure if there's any other detection before this to cancel the update earlier (if nothing changed); also not sure if running `completeUpdate` in the "nothing changed" situation would have side-effects

Reply:
I think the `DeepEqual` is okay. When comparing a task spec to a service spec, there is some subtlety - see `IsTaskDirty`.

Are annotations the only things that can change without causing tasks to become dirty? I would have thought there were other cases, but can't think of any examples offhand. It might make sense to call this something like `serviceSpecChanged` just to be defensive.